File: src/main/java/no/fintlabs/web/archive/CaseSearchParametersService.java
package no.fintlabs.web.archive;

import no.fint.model.felles.basisklasser.Begrep;
import no.fint.model.felles.kompleksedatatyper.Identifikator;
import no.fint.model.resource.arkiv.kodeverk.SaksmappetypeResource;
import no.fint.model.resource.arkiv.kodeverk.TilgangsrestriksjonResource;
import no.fint.model.resource.arkiv.noark.ArkivdelResource;
import no.fint.model.resource.arkiv.noark.KlassifikasjonssystemResource;
import no.fintlabs.cache.FintCache;
import no.fintlabs.model.instance.CaseSearchParametersDto;
import no.fintlabs.model.instance.KlasseDto;
import no.fintlabs.model.instance.SakDto;
import no.fintlabs.model.instance.SkjermingDto;
import org.springframework.stereotype.Service;

import java.util.Optional;
import java.util.StringJoiner;

@Service
public class CaseSearchParametersService {

    private final FintCache<String, ArkivdelResource> arkivdelResourceCache;
    private final FintCache<String, TilgangsrestriksjonResource> tilgangsrestriksjonResourceCache;
    private final FintCache<String, SaksmappetypeResource> saksmappetypeResourceCache;
    private final FintCache<String, KlassifikasjonssystemResource> klassifikasjonssystemResourceCache;

    public CaseSearchParametersService(
            FintCache<String, ArkivdelResource> arkivdelResourceCache,
            FintCache<String, TilgangsrestriksjonResource> tilgangsrestriksjonResourceCache,
            FintCache<String, SaksmappetypeResource> saksmappetypeResourceCache,
            FintCache<String, KlassifikasjonssystemResource> klassifikasjonssystemResourceCache
    ) {
        this.arkivdelResourceCache = arkivdelResourceCache;
        this.tilgangsrestriksjonResourceCache = tilgangsrestriksjonResourceCache;
        this.saksmappetypeResourceCache = saksmappetypeResourceCache;
        this.klassifikasjonssystemResourceCache = klassifikasjonssystemResourceCache;
    }

    public String createFilterQueryParamValue(SakDto sakDto, CaseSearchParametersDto caseSearchParametersDto) {
        StringJoiner filterJoiner = new StringJoiner(" and ");

        if (caseSearchParametersDto.isArkivdel()) {
            sakDto.getArkivdel()
                    .map(arkivdelResourceCache::get)
                    .map(ArkivdelResource::getSystemId)
                    .map(Identifikator::getIdentifikatorverdi)
                    .map(value -> createFilterLine("arkivdel", value))
                    .ifPresent(filterJoiner::add);
        }
        if (caseSearchParametersDto.isTilgangsrestriksjon()) {
            sakDto.getSkjerming()
                    .flatMap(SkjermingDto::getTilgangsrestriksjon)
                    .map(tilgangsrestriksjonResourceCache::get)
                    .map(Begrep::getSystemId)
                    .map(Identifikator::getIdentifikatorverdi)
                    .map(value -> createFilterLine("tilgangskode", value))
                    .ifPresent(filterJoiner::add);
        }
        if (caseSearchParametersDto.isSaksmappetype()) {
            sakDto.getSaksmappetype()
                    .map(saksmappetypeResourceCache::get)
                    .map(Begrep::getSystemId)
                    .map(Identifikator::getIdentifikatorverdi)
                    .map(value -> createFilterLine("saksmappetype", value))
                    .ifPresent(filterJoiner::add);
        }
        if (caseSearchParametersDto.isTittel()) {
            sakDto.getTittel()
                    .map(value -> createFilterLine("tittel", value))
                    .ifPresent(filterJoiner::add);
        }
        if (caseSearchParametersDto.isKlassering()) {
            caseSearchParametersDto.getKlasseringRekkefolge()
                    .map(Integer::parseInt)
                    .ifPresent(rekkefolge -> {

                        Optional<KlasseDto> klasseDto = sakDto.getKlasse()
                                .map(klasseDtos -> klasseDtos.get(rekkefolge));

                        if (caseSearchParametersDto.getKlasseringKlassifikasjonssystem()) {
                            klasseDto
                                    .flatMap(KlasseDto::getKlassifikasjonssystem)
                                    .map(klassifikasjonssystemResourceCache::get)
                                    .map(KlassifikasjonssystemResource::getSystemId)
                                    .map(Identifikator::getIdentifikatorverdi)
                                    .map(value -> createFilterLine(
                                            createKlasseringPrefix(rekkefolge) + "ordning",
                                            value
                                    ))
                                    .ifPresent(filterJoiner::add);
                        }
                        if (caseSearchParametersDto.getKlasseringKlasseId()) {
                            klasseDto
                                    .flatMap(KlasseDto::getKlasseId)
                                    .map(klasseId -> createFilterLine(
                                            createKlasseringPrefix(rekkefolge) + "verdi",
                                            klasseId
                                    ))
                                    .ifPresent(filterJoiner::add);
                        }
                    });
        }
        return filterJoiner.toString();
    }

    private String createFilterLine(String name, String value) {
        return name + " eq '" + value + "'";
    }

    private String createKlasseringPrefix(int rekkefolge) {
        String klassifikasjonName = switch (rekkefolge) {
            case 0 -> "primar";
            case 1 -> "sekundar";
            case 2 -> "tertiar";
            default -> throw new IllegalArgumentException("Rekkefolge must be 0, 1 or 2");
        };
        return "klassifikasjon/" + klassifikasjonName + "/";
    }

}

File: src/main/kotlin/no/novari/flyt/archive/gateway/resource/web/CaseSearchParametersService.kt
package no.novari.flyt.archive.gateway.resource.web

import no.novari.cache.FintCache
import no.novari.fint.model.resource.arkiv.kodeverk.SaksmappetypeResource
import no.novari.fint.model.resource.arkiv.kodeverk.SaksstatusResource
import no.novari.fint.model.resource.arkiv.kodeverk.TilgangsrestriksjonResource
import no.novari.fint.model.resource.arkiv.noark.AdministrativEnhetResource
import no.novari.fint.model.resource.arkiv.noark.ArkivdelResource
import no.novari.fint.model.resource.arkiv.noark.KlassifikasjonssystemResource
import no.novari.flyt.archive.gateway.dispatch.model.instance.CaseSearchParametersDto
import no.novari.flyt.archive.gateway.dispatch.model.instance.KlasseDto
import no.novari.flyt.archive.gateway.dispatch.model.instance.SakDto
import no.novari.flyt.archive.gateway.resource.web.exceptions.KlasseOrderOutOfBoundsException
import no.novari.flyt.archive.gateway.resource.web.exceptions.SearchKlasseOrderNotFoundInCaseException
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service

@Service
class CaseSearchParametersService(
    private val arkivdelResourceCache: FintCache<String, ArkivdelResource>,
    private val administrativEnhetResourceCache: FintCache<String, AdministrativEnhetResource>,
    private val tilgangsrestriksjonResourceCache: FintCache<String, TilgangsrestriksjonResource>,
    private val saksmappetypeResourceCache: FintCache<String, SaksmappetypeResource>,
    private val saksstatusResourceCache: FintCache<String, SaksstatusResource>,
    private val klassifikasjonssystemResourceCache: FintCache<String, KlassifikasjonssystemResource>,
) {
    fun createFilterQueryParamValue(
        sakDto: SakDto,
        caseSearchParametersDto: CaseSearchParametersDto,
    ): String {
        val filterParts = mutableListOf<String>()

        if (caseSearchParametersDto.arkivdel) {
            sakDto.arkivdel
                ?.let(arkivdelResourceCache::get)
                ?.systemId
                ?.identifikatorverdi
                ?.let { createFilterLine("arkivdel", it) }
                ?.let(filterParts::add)
        }
        if (caseSearchParametersDto.administrativEnhet) {
            sakDto.administrativEnhet
                ?.let(administrativEnhetResourceCache::get)
                ?.systemId
                ?.identifikatorverdi
                ?.let { createFilterLine("administrativenhet", it) }
                ?.let(filterParts::add)
        }
        if (caseSearchParametersDto.tilgangsrestriksjon) {
            sakDto.skjerming
                ?.tilgangsrestriksjon
                ?.let(tilgangsrestriksjonResourceCache::get)
                ?.systemId
                ?.identifikatorverdi
                ?.let { createFilterLine("tilgangskode", it) }
                ?.let(filterParts::add)
        }
        if (caseSearchParametersDto.saksmappetype) {
            sakDto.saksmappetype
                ?.let(saksmappetypeResourceCache::get)
                ?.systemId
                ?.identifikatorverdi
                ?.let { createFilterLine("saksmappetype", it) }
                ?.let(filterParts::add)
        }
        if (caseSearchParametersDto.saksstatus) {
            sakDto.saksstatus
                ?.let(saksstatusResourceCache::get)
                ?.systemId
                ?.identifikatorverdi
                ?.let { createFilterLine("saksstatus", it) }
                ?.let(filterParts::add)
        }
        if (caseSearchParametersDto.tittel) {
            sakDto.tittel
                ?.let { createFilterLine("tittel", it) }
                ?.let(filterParts::add)
        }
        if (caseSearchParametersDto.klassering) {
            caseSearchParametersDto.klasseringRekkefolge
                ?.toIntOrNull()
                ?.let { rekkefolge ->
                    val klasseDtoMatchingRekkefolge =
                        sakDto.klasse
                            ?.firstOrNull { klasseDto -> klasseDto.rekkefolge == rekkefolge }
                            ?: throw SearchKlasseOrderNotFoundInCaseException(
                                sakDto.klasse.orEmpty().mapNotNull(KlasseDto::rekkefolge),
                                rekkefolge,
                            )

                    klasseDtoMatchingRekkefolge.rekkefolge?.let { log.debug("KlasseDto rekkefolge: {}", it) }
                    log.debug("Søkeparametere rekkefølge: {}", rekkefolge)

                    if (caseSearchParametersDto.getKlasseringKlassifikasjonssystem()) {
                        klasseDtoMatchingRekkefolge.klassifikasjonssystem
                            ?.let(klassifikasjonssystemResourceCache::get)
                            ?.systemId
                            ?.identifikatorverdi
                            ?.let { value ->
                                createFilterLine(
                                    "${createKlasseringPrefix(rekkefolge)}ordning",
                                    value,
                                )
                            }?.let(filterParts::add)
                    }
                    if (caseSearchParametersDto.getKlasseringKlasseId()) {
                        klasseDtoMatchingRekkefolge.klasseId
                            ?.let { klasseId ->
                                createFilterLine(
                                    "${createKlasseringPrefix(rekkefolge)}verdi",
                                    klasseId,
                                )
                            }?.let(filterParts::add)
                    }
                }
        }

        return filterParts.joinToString(" and ")
    }

    private fun createFilterLine(
        name: String,
        value: String,
    ): String = "$name eq '$value'"

    private fun createKlasseringPrefix(rekkefolge: Int): String {
        val klassifikasjonName =
            when(rekkefolge) {
                1 -> "primar"
                2 -> "sekundar"
                3 -> "tertiar"
                else -> throw KlasseOrderOutOfBoundsException(rekkefolge)
            }

        return "klassifikasjon/$klassifikasjonName/"
    }

    companion object {
        private val log = LoggerFactory.getLogger(CaseSearchParametersService::class.java)
    }
}
